### PR TITLE
feat(entities): add entities actions

### PR DESCRIPTION
**Pull request by 5im0n**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/ngneat/elf/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

First attempt to add Entities Action in Elf, related to the issue: https://github.com/ngneat/elf/issues/247

Issue Number: https://github.com/ngneat/elf/issues/247

## Does this PR introduce a breaking change?

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: packages/entities/src/lib/add.mutation.ts [hunk 1]
```diff
@@ -36,7 +37,7 @@ export function addEntities<
   entities: OrArray<getEntityType<S, Ref>>,
   options: AddEntitiesOptions & BaseEntityOptions<Ref> = {}
 ): Reducer<S> {
-  return function (state, context) {
+  return function (state, context, action) {
```
Comment (NetanelBasal) [inline packages/entities/src/lib/add.mutation.ts:40]: I prefer to put the action on the context and name it `actions`

File: packages/entities/src/lib/entity-actions.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+export enum EntityActions {
```
Comment (NetanelBasal) [inline packages/entities/src/lib/entity-actions.ts:1]: You should export the public apis via the index file

File: packages/entities/src/lib/update.mutation.ts [hunk 3]
```diff
@@ -47,20 +48,27 @@ export function updateEntities<
...
+    const coerceIds = coerceArray(ids);
+    for (const id of coerceIds) {
       updatedEntities[id] = toModel<getEntityType<S, Ref>>(
         updater,
         state[entitiesKey][id]
       );
     }
 
+    // TODO @NetanelBasal
+    // Should we check if entity exist and have been realy update
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:67]: no

File: packages/entities/src/lib/update.mutation.ts [hunk 4]
```diff
@@ -47,20 +48,27 @@ export function updateEntities<
...
+    if (coerceIds.length) {
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:68]: you can move this check to the top and return the original state if the ids array is empty

File: packages/entities/src/lib/update.mutation.ts [hunk 5]
```diff
@@ -318,6 +334,10 @@ export function updateEntitiesIds<
       }
     }
 
+    // TODO @NetanelBasal
+    // Should we propagate the old or the new ids
+    // action.next({ type: EntityActions.Update, ids });
```
Comment (NetanelBasal) [inline packages/entities/src/lib/update.mutation.ts:339]: Always the new for now

File: packages/store/src/lib/operators.ts [hunk 6]
```diff
@@ -51,3 +53,24 @@ export function filterNil<T>(): OperatorFunction<T, NonNullable<T>> {
     (value: T): value is NonNullable<T> => value !== null && value !== undefined
   );
 }
+
+// TODO infer the IDType
+type IDType = string | number;
+
+export function ofTypes<T extends EntityAction<IDType>>(
```
Comment (NetanelBasal) [inline packages/store/src/lib/operators.ts:60]:
```suggestion
export function ofType<T extends EntityAction<IDType>>(
```

File: packages/store/src/lib/operators.ts [hunk 7]
```diff
@@ -51,3 +53,24 @@ export function filterNil<T>(): OperatorFunction<T, NonNullable<T>> {
...
+  }
+
+  const transform = Array.isArray(actionOrActions)
+    ? (action: EntityAction<IDType>) => action
+    : ({ ids }: EntityAction<IDType>) => ids;
+  const actions = coerceArray(actionOrActions);
+
+  return pipe(
+    filter(({ type }) => actions.includes(type)),
+    map((action) => transform(action))
```
Comment (NetanelBasal) [inline packages/store/src/lib/operators.ts:74]:
```suggestion
    map(transform)
```

File: packages/store/src/lib/operators.ts [hunk 8]
```diff
@@ -51,3 +53,24 @@ export function filterNil<T>(): OperatorFunction<T, NonNullable<T>> {
     (value: T): value is NonNullable<T> => value !== null && value !== undefined
   );
 }
+
+// TODO infer the IDType
+type IDType = string | number;
+
+export function ofTypes<T extends EntityAction<IDType>>(
+  actionOrActions?: EntityActions | EntityActions[]
```
Comment (NetanelBasal) [inline packages/store/src/lib/operators.ts:61]:
```suggestion
  actions?: OrArray<EntityAction>
```

File: packages/store/src/lib/operators.ts [hunk 9]
```diff
@@ -51,3 +53,24 @@ export function filterNil<T>(): OperatorFunction<T, NonNullable<T>> {
...
+): OperatorFunction<T, IDType[] | EntityAction<IDType>> {
+  if (isUndefined(actionOrActions)) {
```
Comment (NetanelBasal) [inline packages/store/src/lib/operators.ts:63]: why do we need this condition?

File: packages/store/src/lib/operators.ts [hunk 10]
```diff
@@ -51,3 +53,24 @@ export function filterNil<T>(): OperatorFunction<T, NonNullable<T>> {
...
+    return map((action) => action);
+  }
+
+  const transform = Array.isArray(actionOrActions)
```
Comment (NetanelBasal) [inline packages/store/src/lib/operators.ts:67]: I'd always return the whole action object. In the future it main contain additional properties

File: packages/store/src/lib/store.ts [hunk 11]
```diff
@@ -12,6 +13,9 @@ export class Store<
   state!: State;
   private batchInProgress = false;
 
+  // TODO Infer the EntityAction type
+  private entityActions = new Subject<EntityAction<string | number>>();
```
Comment (NetanelBasal) [inline packages/store/src/lib/store.ts:17]: it should be actions in general. There is no coupling to entities.

```suggestion
  private actions = new Subject<Action<string | number>>();
```

File: packages/store/src/lib/store.ts [hunk 12]
```diff
@@ -23,6 +27,11 @@ export class Store<
     addStore(this);
   }
 
+  // TODO Infer the EntityAction type
+  get actions$(): Observable<EntityAction<string | number>> {
```
Comment (NetanelBasal) [inline packages/store/src/lib/store.ts:31]: there is no need to create a new observable each time, please use a property.

**Review by NetanelBasal — CHANGES_REQUESTED** [review]
Please add specs.

File: packages/entities/src/lib/add.mutation.ts (hunk 1 above)
Comment (5im0n) [reply packages/entities/src/lib/add.mutation.ts:40]: To understand when you say on the context: 

You need before:

```ts
return function (state, actions, context) {
```

or put the actions inside the context

```ts
export type ReducerContext = { config: Record<PropertyKey, any>, actions: EntityAction$ };
```

Comment (NetanelBasal) [reply packages/entities/src/lib/add.mutation.ts:40]: export type ReducerContext = { config: Record<PropertyKey, any>, actions: EntityAction$ };

File: packages/store/src/lib/store.ts (hunk 11 above)
Comment (5im0n) [reply packages/store/src/lib/store.ts]: So the Action is carried in the `store` package ?

However the actions `add`, `set`, `remove` are related to the `entites `package.

```ts
export enum Actions {
  Set = 'Set',
  Add = 'Add',
  Update = 'Update',
  Remove = 'Remove',
}

export interface Action<IDType> {
  type: Actions;
  ids: IDType[];
}
```

Comment (NetanelBasal) [reply packages/store/src/lib/store.ts]: type is a general string.

**Comment by NetanelBasal** [comment]
Let me know when I can recheck it, please.

**Comment by NetanelBasal** [comment]
@5im0n what's the status?

**Comment by mehmet-erim** [comment]
Is there any progress? @5im0n

**Comment by NetanelBasal** [comment]
You have TODOS in the code. No rush. I just wanted to know the status.

**Pull request closed without merging**